packages/plugin-deduplicate/sources/commands/deduplicate.ts

```diff
@@ -0,0 +1,78 @@
+import {BaseCommand}                                    from '@berry/cli';
+import {Configuration, LocatorHash, Project, IdentHash} from '@berry/core';
+import {structUtils}                                    from '@berry/core';
+import {Command}                                        from 'clipanion';
+import * as semver                                      from 'semver';
+
+
+// eslint-disable-next-line arca/no-default-export
+export default class DeduplicateCommand extends BaseCommand {
+  static usage = Command.Usage({
+    category: `Workspace-related commands`,
+    description: `Reduces dependencies with overlapping ranges to a minimal set of packages`,
+    details: `https://github.com/atlassian/yarn-deduplicate for yarn v2`,
+    examples: [],
+  });
+
+  @Command.Path(`deduplicate`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+    const {project} = await Project.find(configuration, this.context.cwd);
+
+    const locatorsByIdent: Map<IdentHash, Set<LocatorHash>> = new Map();
+    for (const [descriptorHash, locatorHash] of project.storedResolutions.entries()) {
+      const value = locatorHash;
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const key = descriptor.identHash;
+
+      const descriptors = locatorsByIdent.get(key);
+      if (descriptors === undefined) {
+        locatorsByIdent.set(key, new Set([value]));
+      } else {
+        locatorsByIdent.set(key, descriptors.add(value));
+      }
+    }
+
+    for (const descriptorHash of project.storedResolutions.keys()) {
+      const descriptor = project.storedDescriptors.get(descriptorHash)!
+      const locatorHashes = locatorsByIdent.get(descriptor.identHash)!;
+
+      if (locatorHashes !== undefined && locatorHashes.size > 1) {
+        const candidates = Array.from(locatorHashes).filter(candidateHash => {
+          const pkg  = project.storedPackages.get(candidateHash)!
+
+          if (structUtils.isVirtualLocator(pkg))
+            return false
+
+          if (pkg.version === null)
+            return false
+
+          const semverMatch = descriptor.range.match(/^npm:(.*)$/);
+          if (semverMatch === null)
+            return false;
```

Comment:
This can be moved outside the function call since you're testing the descriptor, not the locator

Reply:
Yeah and I guess I have to check if the package even satisfies the protocol. Not sure what happens (or even should happen) if a package is resolved with different protocols.